### PR TITLE
:sparkles: markdown and readme api

### PR DESCRIPTION
**Pull request by bpetetot**
#60

### CODE REVIEW

**Comment by bpetetot** [comment]
**New APIs :**
- **`GET /api/readme` :**  retrieve current README.md file content of the component _(if exists)_
- **`POST /api/readme` :** create / save the README.md file of the component with given content
- **`POST /api/markdown`:** generate the markdown from given component docgen

**The goal of these APIs is to have the following process _(in a next PR)_ :** 
1.  The user will click on a button `documentation`
2. It calls **`GET /api/readme`**
     - if _README.md_ already exists => display existing markdown in a markdown editor
     - else call **`POST /api/markdown`** => put generated markdown in a markdown editor
3. The user can update the component documentation through the editor
     - a `save` button will call **`POST /api/readme`** and save / create the README.md file

**I would like also to have buttons in the markdown editor to add some generated parts _(in a next PR)_ :** 
- `Add description` : will add the description where the cursor is in the editor
- `Add props`: will add the component props documentation where the cursor is in the editor
- `Add screenshot` : will add a component screenshot where the cursor is in the editor

To make the `Add description` and `Add props` buttons, I will need to improve API **`POST /api/markdown`** adding some query parameters : 
**`POST /api/markdown?q=props`**

**Comment by fabienjuif** [comment]
I asking myself about endpoints names.
Why don't we use:
 -  **```/api/file```** (to replace **`/api/makrdown`**), and make it more generic
    - **`POST /api/file/readme.md`** by example
 - **```GET /api/documentation```** to generate the documentation, 
    - **`GET /API/documentation?format=markdown|text|html`**

The endpoints don't have to be about format but about their purpose don't you think ?

File: src/script/server/api/services/markdown/docgen2markdown.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
+const generateTitle = (name) => {
```
Comment (fabienjuif) [inline src/script/server/api/services/markdown/docgen2markdown.js]: Why don't we do it in a oneliner ?

File: src/script/server/api/services/markdown/docgen2markdown.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  return `# ${name}\n\n`
```
Comment (fabienjuif) [inline src/script/server/api/services/markdown/docgen2markdown.js]: I think the \n\n shouldn't be there but into the default exports.

File: src/script/server/api/services/markdown/docgen2markdown.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    return Object.keys(props)
+      .reduce((a, b) => {
+        return `${a}${generateProp(b, props[b])}`
+      }, '')
+  }
+  return ''
+}
+
+module.exports = (name, doc) => {
+  let markdown = ''
```
Comment (fabienjuif) [inline src/script/server/api/services/markdown/docgen2markdown.js]: If we delete `\n\n` ending from other functions we could do : 
```es6
module.exports = (name, doc) => `
${generateTitle(name)}

${generateDescription(doc.description)}
`

**Review by fabienjuif — CHANGES_REQUESTED** [review]
Great work, once again :)

I start arguing about endpoints

**Comment by bpetetot** [comment]
Good idea !
We can also merge **`/api/file`** with the current **`/api/fs`**
_Examples :_
- **`GET /api/fs/foo/bar`** will list folder `/foo/bar`
- **`GET /api/fs/foo/bar/README.md`** will get README.md file content
- **`POST /api/fs/foo/bar/README.md`** will create / update README.md file
- ...

And I'm agree about being agnostic about the format with API documentation

**Comment by fabienjuif** [comment]
Yeah I agree about merging both endpoints

**Comment by bpetetot** [comment]
@fabienjuif you can check it

File: src/script/server/api/services/fs/service.js [hunk 4]
```diff
@@ -1,29 +1,45 @@
 const fs = require('fs')
 const path = require('path')
 
+const statFile = (fullpath, stats, content) => Object.assign(
```
Comment (fabienjuif) [inline src/script/server/api/services/fs/service.js]: You can merge master then use spread operator

**Review by fabienjuif — APPROVED** [review]
👌 you rock

**Pull request merged**